src/github.com/getlantern/detour/ioloop.go

```diff
@@ -0,0 +1,348 @@
+package detour
+
+import (
+	"bytes"
+	"io"
+	"sync/atomic"
+)
+
+// to pass result of read operation to merger from underlie connection
+type innerReadResult struct {
+	// the connection which performs this read
+	c conn
+	// buffer holding received data, it's not necessarily the same
+	// buffer as innerReadRequest.
+	buf []byte
+	// IO error, if any
+	err error
+}
+
+type innerReadRequest struct {
+	buf      []byte
+	chResult chan innerReadResult
+}
+
+// ioLoop is the core of detour. It waits for connections and handles
+// read/write requests
+func (dc *Conn) ioLoop() {
+	// Use buffered channel in same goroutine so we can easily add / remove
+	// connections. Should switch to container/ring if performace matters.
+	// It's not safe to access it from multiple goroutines.
+	conns := newConnQueue(2)
+	// Requests ioLoop to remove already closed connections.
+	chRemoveConn := make(chan conn)
+	// Hold the read request so we can re-read after replay.
+	var firstReadReq *innerReadRequest
+
+	// Can we replay this request
+	// TODO: How to check when using pipelining or HTTP/2?
+	var nonidempotentHTTPRequest bool
+
+	for {
+		select {
```

Comment:
After Lantern running for hours, 100+ goroutines blocked here, which means the caller doesn't close `detour.Conn` at all! Let's see how many goroutines will accumulate after running for days.